nixos/modules/programs/steam.nix

```diff
@@ -148,6 +156,11 @@ in {
         allowedTCPPorts = [ 27015 ]; # SRCDS Rcon port
         allowedUDPPorts = [ 27015 ]; # Gameplay traffic
       })
+
```

Comment:
The docs from steam say that 27031-37036 is required for LAN discovery:

https://help.steampowered.com/en/faqs/view/46BD-6BA8-B012-CE43

Since this overlaps with the LAN discovery used for remote play, I think it'd make sense just to separate it into a different `mkIf` block:

```diff
--- a/nixos/modules/programs/steam.nix
+++ b/nixos/modules/programs/steam.nix
@@ -66,6 +66,14 @@ in {
       '';
     };
 
+    localNetworkTransfers.openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Open ports in the firewall for Steam Local Network Game Transfers.
+      '';
+    };
+
     remotePlay.openFirewall = mkOption {
       type = types.bool;
       default = false;
@@ -139,8 +147,16 @@ in {
     ] ++ lib.optional cfg.gamescopeSession.enable steam-gamescope;
 
     networking.firewall = lib.mkMerge [
+      (mkIf cfg.localNetworkTransfers.openFirewall {
+        allowedTCPPorts = [ 27040 ];
+      })
+
       (mkIf cfg.remotePlay.openFirewall {
         allowedTCPPorts = [ 27036 ];
+      })
+
+      (mkIf (cfg.localNetworkTransfers.openFirewall || cfg.remotePlay.openFirewall) {
+        # LAN client discovery
         allowedUDPPortRanges = [ { from = 27031; to = 27036; } ];
       })
```

Reply:
I'm not really sure what the convention here is, but my gut is telling me that if less ports need to be open for something to function, then less ports should be open. If the used port change, this can be updated when they do.

If the convention is to follow documentation, then so be it, but I will note that in this instance the documentation was wrong in two different ways: swapped digits 4 and 7, and missing UDP ports, meaning following the docs as they were wouldn't have made this function.

The only reason I know which ports are used in practice is because I checked; the docs were updated following my question to martino, after I had already made the PR.

As for the patch: I do like splitting LAN discovery into its own section. Makes the overlap more obvious.

Reply:
Oh sorry! Yeah I missed your mention of the docs originally being wrong in the PR summary. I was looking into this myself before seeing this PR. I saw the right information there, but didn't realize it was fixed after you reached out 😅.

Yeah that's a good gut reaction. It really doesn't make much sense why they'd need the full range. I assume steam tries to scan it as a fallback in case one of them fails?

I was mainly just thinking that since remote play also uses the same UDP port range, it'd be nice to group them together as generic "lan discovery ports". Maybe we should also reduce remote play to only use one UDP port too?

Reply:
We could limit Remote Play ports, obviously assuming it doesn't use those to stream the data and break. I haven't looked into how it works though, and that might be better left for another PR anyway. :P I'd hope they had a reason for listing the other ports as being used by Remote Play though.

I have a mild hunch that since local network transfers use discovery infrastructure used by Remote Play, they might've blindly copied all UDP ports used by Remote Play without fully understanding what they're used for, but that might not be giving them enough credit.

I'm also wondering what ports the Steam process keeps open. It might be a useful clue.